src/python/pants/engine/rules.py

```diff
@@ -350,6 +279,28 @@ def _uncacheable_rule(*args, **kwargs) -> Callable:
     return inner_rule(*args, **kwargs, rule_type=RuleType.uncacheable_rule, cacheable=False)
 
 
+def rule_helper(func: Callable) -> Callable:
+    """Decorator which marks a static function as a "rule helper".
+
+    Functions marked as rule helpers are allowed to be called by rules and other rule helpers
+    and can `await Get/MultiGet`. The rule parser adds these functions' awaitables to the rule's
+    awaitables.
+
+    There are a few restrictions:
+        - Rule helpers must be "private". I.e. start with an underscore
+        - Rule helpers must be called by accessing module-scoped attributes
```

Comment:
Yeah this one is hard to explain. Basically the start of any attribute chain leading to a rule helper must be at module scope.

Eg `foo.bar.baz` or `foo()` must have `foo` at module scope.

Additionally it means this is invalid `foo().bar` because it's not an attribute chain.

Happy to hear suggestions for wording

Reply:
Sorry, I still don't think I understand. Is this something to do with only being able to import helpers from the same module?

Reply:
I think this means that as long as you refer to helpers and types you import at the top level (i.e. not in a nested scope) you're good. And also, that any helpers defined in classes, you need to have an instance of that class, or access a static/class helper rather than creating a new instance in-place.

```python

from my.helpers import util_rules

@rule
def example() -> str:
  # this should be ok, attribute chain rooted at module level
  util_rules._some_helper(arg1, arg2)

  # not ok, rooted in a nested scope
  from some.where import other
  other._some_helper(arg1, arg2)

```